A39-xds-http-filters.md

```diff
@@ -0,0 +1,268 @@
+  `type_url` field) that is present in the filter registry.  Note that the
+  set of known filter config types will be different on the gRPC client and
+  server; if the `HttpConnectionManager` proto is inside an HTTP API Listener,
+  it will look only at filters registered for the gRPC client, whereas if it
+  is inside a TCP Listener, it will look only at filters registered for
+  the gRPC server.  Any unknown config type will cause the Listener
+  resource to be NACKed, unless the `is_optional` field is true, in which
+  case the individual filter will be ignored.
+- The fields in the filter config will be validated by the filter
+  implementation.  Any validation error will cause the Listener resource
+  to be NACKed.
+
+#### Filter Config Overrides
+
+In each of the `VirtualHost`, `Route`, and `WeightedCluster.ClusterWeight`
+protos, there is a field called `typed_per_filter_config` that contains
+per-filter overrides.  Each entry in the map will be used as follows:
+- The entry's key is the filter instance name to which the config override
+  applies.  The entry will be used for the filter in the
+  `HttpConnectionManager.http_filters` list with the same `name` field.
+- The entry's value contains the override config for the filter.  As with
+  the `HttpConnectionManager.http_filters.typed_config` field, the type
+  of the protobuf message determines which filter is used, based on the
+  filters known to the filter registry.
+- If the value protobuf message type is `envoy.config.route.v3.FilterConfig`,
+  then gRPC will look inside that type's
+  [`config`](https://github.com/envoyproxy/envoy/blob/9ed95162ce11a09a56c1743d0a6e342d9bbfd2b6/api/envoy/config/route/v3/route_components.proto#L1966)
+  field to determine the actual type.  In addition, gRPC will look at the
+  [`is_optional`](https://github.com/envoyproxy/envoy/blob/9ed95162ce11a09a56c1743d0a6e342d9bbfd2b6/api/envoy/config/route/v3/route_components.proto#L1971)
+  field to determine how to handle unknown protobuf message types.
+- Whether or not the `FilterConfig` wrapper is used, if the type is
+  `udpa.type.v1.TypedStruct`, then gRPC will look inside that type's
+  [`type_url`](https://github.com/cncf/udpa/blob/cc1b757b3eddccaaaf0743cbb107742bb7e3ee4f/udpa/type/v1/typed_struct.proto#L38)
+  field to determine the actual type; this allows for third-party
+  plugins to encode their configurations as a `google.protobuf.Struct`
+  proto instead of using a concrete proto message type that the client
+  may not be able to handle.
+
+Note that Envoy currently keys these maps by legacy filter names instead
+of by filter instance names, but this is expected to change, as per
+https://github.com/envoyproxy/envoy/issues/12274.  gRPC will directly
+implement the desired semantics here.
+
+When gRPC receives the `RouteConfiguration` proto, either in the LDS or
+RDS response, it will validate the contents of each
+`typed_per_filter_config` map as follows:
```

Comment:
Where's the stuff about "fail validation if using a config in the wrong context (listener filter list vs override config)" that was discussed earlier today? If that's implied by "The fields in the filter config should be validated by the filter implementation", then that needs to be strongly clarified. This applies to `HttpConnectionManager.http_filters` as well.

Reply:
Clarified this in both places.